File: sdk/js/wasm-sdk.js
// @mallchain/wasm-sdk - JavaScript SDK for Mallchain Smart Contracts
// 
// npm install @mallchain/wasm-sdk
// 
// Usage:
// import { MallchainWasmClient } from '@mallchain/wasm-sdk'
// const client = new MallchainWasmClient('https://rpc.mallchain.com')

export interface ContractExecution {
  code_id?: number
  contract_address?: string
  sender: string
  msg: Record<string, unknown>
  label?: string
  init_msg?: Record<string, unknown>
}

export interface ContractResponse<T = unknown> {
  data?: T
  events?: ContractEvent[]
  gas_used?: number
}

export interface ContractEvent {
  type: string
  attributes: Record<string, string>
}

export interface WasmClientOptions {
  rpcUrl: string
  chainId?: string
}

export class MallchainWasmClient {
  private rpcUrl: string
  private chainId: string
  private txEncoder: (msg: unknown) => Promise<string>

  constructor(options: WasmClientOptions) {
    this.rpcUrl = options.rpcUrl
    this.chainId = options.chainId || 'mall-1'
  }

  async storeCode(wasmBytecode: Uint8Array, sender: string): Promise<number> {
    const code = await this.encodeAndBroadcast('wasm', 'store', {
      wasm_code: Buffer.from(wasmBytecode).toString('base64'),
      sender
    })
    return code.code_id
  }

  async instantiate(codeId: number, initMsg: Record<string, unknown>, sender: string, label?: string): Promise<string> {
    const contract = await this.encodeAndBroadcast('wasm', 'instantiate', {
      code_id: codeId,
      sender,
      label,
      init_msg: initMsg
    })
    return contract.contract_address
  }

  async execute(contractAddress: string, msg: Record<string, unknown>, sender: string, gasLimit?: number): Promise<ContractResponse> {
    const result = await this.encodeAndBroadcast('wasm', 'execute', {
      contract_address: contractAddress,
      sender,
      msg,
      gas_limit: gasLimit
    })
    return result
  }

  async query(contractAddress: string, query: Record<string, unknown>): Promise<ContractResponse> {
    // Query via REST endpoint
    const res = await fetch(`${this.rpcUrl}/wasm/contract/${contractAddress}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ query })
    })
    return res.json()
  }

  private async encodeAndBroadcast(module: string, action: string, msg: unknown): Promise<unknown> {
    // Placeholder - would integrate with CosmJS or @cosmos-kit
    throw new Error(`Not implemented - requires CosmJS integration for ${module} ${action}`)
  }
}

// Example contract interaction:
// const client = new MallchainWasmClient({ rpcUrl: 'https://rpc.mallchain.com' })
// const codeId = await client.storeCode(wasmBuffer, signerAddress)
// const contractAddr = await client.instantiate(codeId, { name: 'MyContract' }, signerAddress, 'my-contract')
// const result = await client.execute(contractAddr, { action: 'transfer', to: 'mall1...', amount: 100n }, signerAddress)

File: sdk/python/wasm_sdk.py
# mallchain-wasm-sdk - Python SDK for Mallchain Smart Contracts
#
# pip install mallchain-wasm-sdk
#
# Usage:
# from mallchain_wasm_sdk import WasmClient
# client = WasmClient('https://rpc.mallchain.com')
# code_id = client.store_code(wasm_bytes, sender)
# contract_addr = client.instantiate(code_id, init_msg, sender, label='my-contract')

from typing import Any, Optional
import json
import requests


class WasmClient:
    def __init__(self, rpc_url: str, chain_id: str = "mall-1"):
        self.rpc_url = rpc_url
        self.chain_id = chain_id

    def store_code(self, wasm_bytes: bytes, sender: str) -> int:
        """Store WASM bytecode and return code ID"""
        msg = {
            "type": "wasm/MsgStoreCode",
            "value": {
                "wasm_code": wasm_bytes.hex(),
                "sender": sender
            }
        }
        # Requires signing integration
        raise NotImplementedError("Requires CosmJS integration")

    def instantiate(
        self,
        code_id: int,
        init_msg: dict,
        sender: str,
        label: Optional[str] = None
    ) -> str:
        """Instantiate a WASM contract and return contract address"""
        msg = {
            "type": "wasm/MsgInstantiateContract",
            "value": {
                "code_id": code_id,
                "sender": sender,
                "label": label,
                "init_msg": init_msg
            }
        }
        raise NotImplementedError("Requires signing integration")

    def execute(
        self,
        contract_addr: str,
        msg: dict,
        sender: str,
        gas_limit: int = 200000
    ) -> dict:
        """Execute a WASM contract"""
        response = requests.post(
            f"{self.rpc_url}/wasm/execute/{contract_addr}",
            json={"msg": msg, "sender": sender, "gas_limit": gas_limit}
        )
        return response.json()

    def query(self, contract_addr: str, query: dict) -> Any:
        """Query a WASM contract"""
        response = requests.post(
            f"{self.rpc_url}/wasm/contract/{contract_addr}",
            json={"query": query}
        )
        return response.json()


# Example usage:
# client = WasmClient('https://rpc.mallchain.com')
# # code_id = client.store_code(wasm_buffer, signer_address)
# # contract_addr = client.instantiate(code_id, {'name': 'MyContract'}, signer_address, 'rewards')
# # result = client.execute(contract_addr, {'action': 'transfer', 'to': 'mall1...', 'amount': 100}, signer_address)